packages/ses/src/reporting.js

```diff
@@ -0,0 +1,104 @@
+import { globalThis, TypeError } from './commons.js';
+import { assert } from './error/assert.js';
+
+/**
+ * @import {Reporter, GroupReporter} from './reporting-types.js'
+ */
+
+/**
+ * Creates a suitable reporter for internal errors and warnings out of the
+ * Node.js console.error to ensure all messages to go stderr, including the
+ * group label.
+ * Accounts for the extra space introduced by console.error as a delimiter
+ * between the indent and subsequent arguments.
+ *
+ * @param {(...message: Array<any>) => void} print
+ */
+const makeReportPrinter = print => {
+  let indent = false;
+  /** @param {Array<any>} args */
+  const printIndent = (...args) => {
+    if (indent) {
+      print(' ', ...args);
+    } else {
+      print(...args);
+    }
+  };
+  return /** @type {GroupReporter} */ ({
+    warn(...args) {
+      printIndent(...args);
+    },
+    error(...args) {
+      printIndent(...args);
+    },
+    groupCollapsed(...args) {
+      assert(!indent);
+      print(...args);
+      indent = true;
+    },
+    groupEnd() {
+      indent = false;
+    },
+  });
+};
+
+const mute = () => {};
+
+/**
+ * @param {"platform" | "console" | "none"} reporting
+ */
+export const chooseReporter = reporting => {
+  if (reporting === 'none') {
+    return makeReportPrinter(mute);
+  }
+  if (reporting !== 'platform' && reporting !== 'console') {
+    throw new TypeError(`Invalid lockdown reporting option: ${reporting}`);
+  }
+  if (
+    reporting === 'console' ||
+    globalThis.window === globalThis ||
+    globalThis.importScripts !== undefined
+  ) {
+    return console;
+  }
+  if (globalThis.console !== undefined) {
+    // On Node.js, we send all feedback to stderr, regardless of purported level.
+    return makeReportPrinter(globalThis.console.error);
+  }
+  if (globalThis.print !== undefined) {
+    return makeReportPrinter(globalThis.print);
+  }
+  return makeReportPrinter(mute);
+};
+
+/**
+ * @param {string} groupLabel
+ * @param {GroupReporter} console
+ * @param {(internalConsole: Reporter) => void} callback
+ */
+export const reportInGroup = (groupLabel, console, callback) => {
```

Comment:
While I would agree it is overkill to assert that the `groupLabel` is the same every time, if you don't assert it, you should at least mention in the doc-comment that only the first `groupLabel` in a group matters, and any inconsistency between them is ignored.

Reply:
I don’t grok this comment, but I can more rigorously describe `reportInGroup` above. Suffice it to say that it does not compose better than the underlying `console` group behavior and is very sensitive to context. We only ever have one reporter active on the stack at any given time, either on the stack of `repairIntrinsics` or `hardenIntrinsics` (for enablements).

Reply:
I was indeed confused. Resolving.